### ISSUE TITLE
Compression based side-channel attacks and BREACH

### ISSUE THREAD

**Issue by randomstuff** [OWASP/ASVS#2203]
Should we include requirements for compression based side-channel attacks in general and BREACH (which is still alive and well as far as I know) in particular?

Options:

* general requirement
* specific requirement for BREACH
* both

cf. #1798

**Comment by elarlang** [comment]
ping @Sjord

**Comment by jmanico** [comment]
Only really old ciphers are vulnerable to BREACH. The crypto requirements saying use modern ciphers should cover this problem.

**Comment by randomstuff** [comment]
@jmanico, Do you have a reference on this? The references I found seems to says otherwise (or at least does not suggest that modern crypto mitigates BREACH).

References:

* [Implement TLS 1.3 random record padding to mitigate BREACH](https://trac.nginx.org/nginx/ticket/1977)
* [StackExchange cryptography answer](https://crypto.stackexchange.com/a/108151/2524)

**Comment by jmanico** [comment]
Yep, you’re right and I stand corrected. Here is a more comprehensive resource that discusses how it’s TLS version and cipher agnostic.BREACH ATTACKbreachattack.comDefenses this resource suggests include:Disabling HTTP compression Separating secrets from user input Randomizing secrets per request Masking secrets (effectively randomizing by XORing with a random secret per request) Protecting vulnerable pages with CSRF Length hiding (by adding random number of bytes to the responses) Rate-limiting the requestsMessage ID: ***@***.***>

**Comment by ryarmst** [comment]
I would recommend against a specific requirement for BREACH specifically because there are many similar vectors. As one example (which I don't think has been published on yet - my bad for not getting around to it), web-based PDF generators using text compression can also be utilized for side-channel compression attacks. BREACH could be mitigated by disabling HTTP compression, but this would not mitigate other forms of side-channel compression attacks.

The broad mitigations generally overlap with those to protect against XS-Leaks, but I would not technically classify these as XS-Leaks. I think controls may be most appropriate in V50, but I haven't kept up-to-date on the latest discussions. Ultimately, it requires restricting cross-site/cross-origin requests (not just read capability using CORS, but preventing a response from the server). 

Example mitigations:
- Using Origin (or Fetch Metadata) headers to reject requests from unintended sources
- For apps using cookies for sessions, restricting transmission via the *samesite* attribute
- Anything else specific to the mechanism or business logic..

**Comment by randomstuff** [comment]
> I would recommend against a specific requirement for BREACH specifically because there are many similar vectors.

Yes, at the same time BREACH is a very simple and generic vector with very simple/actionable mitigations ("in doubt just don't enable HTTP compression for dynamic responses in your server").

**Comment by randomstuff** [comment]
What I mean is that having some wording targeting explicitly BREACH would help the users of the ASVS to at least tackle BREACH whereas the general case seems quite hard to grasp and not so easy to mitigate.

So I would say that a general requirement would be necessary but that an additional BREACH-specific requirement would be useful as well (even if it would strictly speaking be redundant).

**Comment by ryarmst** [comment]
I'm not a performance engineer and am not certain of the significance of HTTP level compression, but I suspect a broad requirement to disable compression for dynamic responses may not be universally feasible.

**Comment by randomstuff** [comment]
Yes I did mean to say that the requirement should be "you must disable HTTP compression for dynamic content" :)

**Comment by Sjord** [comment]
> For apps using cookies for sessions, restricting transmission via the samesite attribute

We could have a requirement for SameSite=strict on session cookies under certain conditions. Then an attacker cannot forge cross-site requests that retrieve information from the user's session.

**Comment by randomstuff** [comment]
Could we have something such as:

> Verify that the application mitigates compression based side-channel attacks.

and

> Verify that the application mitigates BREACH attacks if HTTP compression is used in dynamic HTTPS response.

This hints that a mitigation is to disable HTTP compression in dynamic HTTP responses (and that if you use HTTP compression in dynamic HTTP responses you may want to consider implementation another mitigation).

**Comment by ryarmst** [comment]
Thinking about it further, the issue can be generalized to when the following co-occur:
1. Compression of sensitive information occurs *before* encryption.
2. An attacker is either able to inject (chosen) or observe (known) plaintext (the latter is probably unlikely to pose significant risk) and also observe encrypted data (most likely network sniffing).

In addition to BREACH (HTTP compression) and app-specific functionality like PDFs (PDF compression), this is also mentioned in [RFC 8725](https://datatracker.ietf.org/doc/html/rfc8725#name-avoid-compression-of-encryp) for JWTs.

Therefore, perhaps a general requirement to avoid compress-then-encrypt patterns for V6?

**Comment by ryarmst** [comment]
Ping @danielcuthbert

**Comment by randomstuff** [comment]
> Therefore, perhaps a general requirement to avoid compress-then-encrypt patterns for V6?

Yes, it might be useful to have a general requirement in V6. (I am wondering if the uses cases outside of HTTP compression are not somewhat niche though.)

The HTTP compression is probably important-enough to warrant some proper wording or event dedicated requirement outside of V6 for better exposition (?). If BREACH is only covered as a subclause of a general requirement in V6, it might easily be overlooked.

**Comment by tghosth** [comment]
Having discussed it between us, we think that a requirement would be very hard to implement correctly and also potentially have a low impact on security posture as the likelihood of a successful attack is low. We also think that other cross origin controls would hopefully help mitigate this.
